### DISCUSSION TITLE
Omit the required field on upsert in an associated table

### DISCUSSION THREAD

**Discussion by maiconsanson** [prisma/prisma#23796] [Q&A]
### Question

Hi, I'm creating a query using the `update` in the main table with the `upsert` on an associated table using a compound ID.

Even if the `where` clause using the compound ID (`bag_id_locale_id`) results in `true`, **the upsert from the associated table bananas is always calling the create method**, which fails because the `title` is required.

Once the compound ID is found, is there a reason to create a new associated item?
Is there a way to completely omit a required field to update only what I send? Send `undefined` or don't send the field isn't working.

### Expected behavior (optional)

If the `where` of the compound ID `bag_id_locale_id` is `true`, the `created` method shouldn't be called when using `upsert`.

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
model bags {
  id Int @id @default(autoincrement())
  user_id Int
  some_bag_field String? @db.VarChar(255)

  @@index([user_id])

  user users @relation(fields: [user_id], references: [id], onDelete: Cascade)
  bananas bananas[]
}

model bananas {
  id Int @id @default(autoincrement())
  bag_id Int
  locale_id Int
  title String @db.VarChar(255)
  description String? @db.VarChar(500)

  // compound ID
  @@unique([bag_id, locale_id])
  @@index([bag_id])
  @@index([locale_id])

  bag bags @relation(fields: [bag_id], references: [id], onDelete: Cascade)
  locale locales @relation(fields: [locale_id], references: [id], onDelete: NoAction)
}
```

```js
const updatedBag = await prisma.bags.update({
  where: { id: bag_id, user_id },
  data: {
    some_bag_field,

    bananas: {
      upsert: {
        where: {
          // compound ID
          bag_id_locale_id: { bag_id, locale_id },
        },
        create: {
          description,
          locale: { connect: { id: locale_id } },
          title,
        },
        update: {
          description,
          title,
        },
      },
    },
  },
});
```

- OS: **macOS**
- Database: **PostgreSQL**
- Node.js version: **v20.10.0**

- Run `prisma -v` to see your Prisma version and paste it
```
prisma                  : 5.12.1
@prisma/client          : 5.12.1
Computed binaryTarget   : darwin-arm64
Operating System        : darwin
Architecture            : arm64
```

**Comment by ludralph** [comment]
Hi @maiconsanson 👋 

Thank you for raising this question.

The issue is related to how upsert works within a nested update. Here's why your current approach is failing:

The `where` clause in the upsert for the bananas relation checks for the compound ID `(bag_id and locale_id)`. If a record exists, with that compound ID, Prisma Client attempts to use the `update` block within `upsert`. However, since `title` is a required field, if the existing record doesn't have a `title` value, the update fails. As the `update` fails, Prisma Client falls back to the `create` block.
But since you haven't provided a value for title (which is required), the create operation also fails.

To fix this issue, you need to provide a value for `title`. This way, if an update operation is called, the value of `title` will be used during the `update`, preventing the `create` operation from being triggered.

**Reply by maiconsanson** [reply]
Hello @ludralph. Thank you for your response!
I thought an `undefined` would be omitted even if the field is required. 
https://www.prisma.io/docs/orm/prisma-client/special-fields-and-types/null-and-undefined

That's why I'm confused about why the `create` is firing.
Even in the case of a required field, shouldn't the `undefined` value be omitted? Don't send the required field is not working too, the create method is always called.

I've just found a related issue.
https://github.com/prisma/prisma/issues/14190

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
